pkg/controller/master/addon/addon_processing.go

```diff
@@ -0,0 +1,301 @@
+// based on status of job, the addon deployment will be marked successful or errored
+// and status will be updated accordingly
+func (h *Handler) processEnablingAddon(a *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	aObj := a.DeepCopy()
+	hc, _, err := h.getAddonHelmChart(aObj)
+	if err != nil {
+		return aObj, err
+	}
+
+	if hc == nil {
+		// unable to find helmChart. Likely manually GC'd. Reset to initial state
+		// this will allow addon to be processed again and redeployed
+		markCompletedCondition(aObj)
+		aObj.Status.Status = harvesterv1.AddonDisabled
+		return h.addon.UpdateStatus(aObj)
+	}
+
+	logrus.Debugf("checking status of addon: %s", aObj.Name)
+	return h.reconcileCurrentInstallationJob(hc, aObj, harvesterv1.AddonDeployed)
+}
+
+// processAddonUpgrades will check if an update is needed for the helm chart
+// and if needed will update the helmChart object spec
+func (h *Handler) processAddonUpgrades(a *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	aObj := a.DeepCopy()
+
+	hc, _, err := h.getAddonHelmChart(aObj)
+
+	if err != nil {
+		return aObj, err
+	}
+
+	if hc == nil {
+		return aObj, h.deployHelmChart(aObj)
+	}
+	needed, vals, err := h.isHelmchartUpdateNeeded(aObj, hc)
+	if err != nil {
+		return aObj, err
+	}
+
+	if needed {
+		hc, _, err := h.getAddonHelmChart(aObj)
+		if err != nil {
+			return aObj, err
+		}
+		hc.Spec.Chart = aObj.Spec.Chart
+		hc.Spec.Repo = aObj.Spec.Repo
+		hc.Spec.Version = aObj.Spec.Version
+		hc.Spec.ValuesContent = vals
+		_, err = h.helm.Update(hc)
+		if err != nil {
+			return aObj, err
+		}
+
+		if aObj.Status.Status != harvesterv1.AddonUpdating || !harvesterv1.AddonOperationInProgress.IsTrue(aObj) {
+			aObj.Status.Status = harvesterv1.AddonUpdating
+			markInProgressCondition(aObj)
+			return h.addon.UpdateStatus(aObj)
+		}
+	}
+
+	return aObj, nil
+}
+
+// wait for AddonUpgrdes will wait for the helmChart upgrade to be completed
+// and accordingly reconcile state
+func (h *Handler) waitForAddonUpgrades(a *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	aObj := a.DeepCopy()
+
+	hc, _, err := h.getAddonHelmChart(aObj)
+	if err != nil {
+		return aObj, err
+	}
+
+	if hc == nil {
+		// unable to find helmChart. Likely manually GC'd. Reset to initial state
+		// this will allow addon to be processed again and redeployed
+		markCompletedCondition(aObj)
+		aObj.Status.Status = harvesterv1.AddonDisabled
+		return h.addon.UpdateStatus(aObj)
+	}
+
+	return h.reconcileCurrentInstallationJob(hc, aObj, harvesterv1.AddonDeployed)
+}
+
+func (h *Handler) reconcileCurrentInstallationJob(hc *helmv1.HelmChart, aObj *harvesterv1.Addon, state harvesterv1.AddonState) (*harvesterv1.Addon, error) {
+	j, ok, err := h.currentInstallationJob(hc, aObj)
+	if err != nil {
+		return aObj, err
+	}
+
+	// the current job on hc is from the previous deploy/update
+	// or job is not complete. then wait for this to be completed
+	if !ok || !isJobComplete(j) {
```

Comment:
should be && ?

Reply:
it is correct. The idea is to requeue the addon and check again if the following conditions are met:
* job doesnt exist 
* job does exist but is not is not complete.